### PR TITLE
Notifications

### PR DESCRIPTION
**Pull request by kkarkus**

### CODE REVIEW

File: pimcore/models/Notification.php [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+     */
+    public function getId()
+    {
+        return $this->id;
+    }
+
+    /**
+     * @param int $id
+     *
+     * @return Notification
```
Comment (luklewluk) [inline pimcore/models/Notification.php]: Should be `@return $this`

File: pimcore/models/Notification.php [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+     */
+    public function getType()
+    {
+        return $this->type;
+    }
+
+    /**
+     * @param string $type
+     *
+     * @return Notification
```
Comment (luklewluk) [inline pimcore/models/Notification.php]: `@return $this`

File: pimcore/models/Notification.php [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+     */
+    public function getTitle()
+    {
+        return $this->title;
+    }
+
+    /**
+     * @param string $title
+     *
+     * @return Notification
```
Comment (luklewluk) [inline pimcore/models/Notification.php]: `@return $this`

File: pimcore/models/Notification.php [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+     */
+    public function getMessage()
+    {
+        return $this->message;
+    }
+
+    /**
+     * @param string $message
+     *
+     * @return Notification
```
Comment (luklewluk) [inline pimcore/models/Notification.php]: `@return $this`

File: pimcore/models/Notification/Dao.php [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+ * @property \Pimcore\Model\Notification $model
+ */
+class Dao extends Model\Dao\AbstractDao
+{
+    /**
+     * Fetch a row by an id from the database and assign variables to the document model.
+     *
+     * @param $id
+     *
+     * @throws \Exception
```
Comment (luklewluk) [inline pimcore/models/Notification/Dao.php:34]: `@return void` (same all methods below)

File: pimcore/models/Notification/Dao.php [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+                "creationDate" => time()
+            ]);
+
+            $this->model->setId($this->db->lastInsertId());
+        } catch (\Exception $e) {
+            throw $e;
+        }
+    }
+
+
```
Comment (luklewluk) [inline pimcore/models/Notification/Dao.php]: Unnecessary extra line

File: pimcore/models/Notification/Listing.php [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+namespace Pimcore\Model\Notification;
+
+use Pimcore\Model;
+use Pimcore\Model\Notification;
+
+/**
+ * @method int getTotalCount()
+ * @method int getCount()
+ * @method int loadIdList()
+ */
```
Comment (luklewluk) [inline pimcore/models/Notification/Listing.php]: It should be in class description (under `@method ... getDao()` )

File: pimcore/models/Notification/Listing.php [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+ * @author Kamil Karkus <kkarkus@divante.pl>
+ *
+ * @method \Pimcore\Model\Notification\Listing\Dao getDao()
+ */
+class Listing extends Model\Listing\AbstractListing implements \Zend_Paginator_Adapter_Interface, \Zend_Paginator_AdapterAggregate, \Iterator
+{
+    /**
+     * Contains the results of the list
+     *
+     * @var array
```
Comment (luklewluk) [inline pimcore/models/Notification/Listing.php]: I think `@var Notification[]` would be more helpful

File: pimcore/models/Notification/Listing/Dao.php [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+/**
+ * @author Piotr Ćwięcek <pcwiecek@divante.pl>
+ * @author Kamil Karkus <kkarkus@divante.pl>
+ *
+ * @property \Pimcore\Model\Notification\Listing $model
+ */
+class Dao extends Model\Listing\Dao\AbstractDao
+{
+
```
Comment (luklewluk) [inline pimcore/models/Notification/Listing/Dao.php]: Unnecessary extra line

File: pimcore/models/Notification/Listing/Dao.php [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+    /** @var  Callback function */
```
Comment (luklewluk) [inline pimcore/models/Notification/Listing/Dao.php]: Extra space before `Callback`

File: pimcore/models/Notification/Listing/Dao.php [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+
+    /**
+     * @return int
+     */
+    public function getTotalCount()
+    {
+        try {
+            $amount = (int) $this->db->fetchOne("SELECT COUNT(*) as amount FROM notifications " . $this->getCondition(), $this->model->getConditionVariables());
+        } catch (\Exception $e) {
+        }
```
Comment (luklewluk) [inline pimcore/models/Notification/Listing/Dao.php:116]: We should log the exception. If there's no need to log it, it's a good practice to add comment why error can be just skipped.

File: pimcore/modules/admin/controllers/NotificationController.php [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+
+/**
+ * Class Admin_NotificationController
+ *
+ * @author Piotr Ćwięcek <pcwiecek@divante.pl>
+ * @author Kamil Karkus <kkarkus@divante.pl>
+ */
+class Admin_NotificationController extends \Pimcore\Controller\Action\Admin\Document
+{
+
```
Comment (luklewluk) [inline pimcore/modules/admin/controllers/NotificationController.php:25]: Unnecessary extra line

File: pimcore/models/Notification.php [hunk 13]
```diff
@@ -0,0 +1,335 @@
...
+ * LICENSE.md which is distributed with this source code.
+ *
+ * @category   Pimcore
+ * @package    Notification
+ * @copyright  Copyright (c) 2009-2016 pimcore GmbH (http://www.pimcore.org)
+ * @license    http://www.pimcore.org/license     GPLv3 and PEL
+ */
+
+namespace Pimcore\Model;
+use Pimcore\Logger;
```
Comment (luklewluk) [inline pimcore/models/Notification.php:19]: `namespace` and `use` should be separated by extra line

File: pimcore/models/Notification.php [hunk 14]
```diff
@@ -0,0 +1,335 @@
...
+
+    /**
+     * @var string
+     */
+    public $message = '';
+
+    /**
+     * @var null|int
+     */
+    public $fromUser = null;
```
Comment (luklewluk) [inline pimcore/models/Notification.php]: No need to add `= null` as it is a default value anyway

File: pimcore/models/Notification.php [hunk 15]
```diff
@@ -0,0 +1,335 @@
...
+
+    /**
+     * @var int
+     */
+    public $modificationDate;
+
+    /**
+     * @param int $id
+     *
+     * @return Notification
```
Comment (luklewluk) [inline pimcore/models/Notification.php]: Or `null` (line 83)

**Review by luklewluk — COMMENTED** [review]
A few things I would change :)
Also I think it's worth to add a description to every class and method you've created.

File: pimcore/modules/admin/controllers/NotificationController.php (hunk 12 above)
Comment (kkarkus) [reply pimcore/modules/admin/controllers/NotificationController.php:25]: http://www.php-fig.org/psr/psr-2/#2-3-lines

> "Blank lines MAY be added to improve readability and to indicate related blocks of code."

Comment (kkarkus) [reply pimcore/modules/admin/controllers/NotificationController.php:25]: @luklewluk Thank you for comments.

**Pull request merged**